### ISSUE TITLE
Consistency of family()$variance

### ISSUE THREAD

**Issue by kaskr** [glmmTMB/glmmTMB#294]
I think we have some consistency problems which I wasn't really aware of until now.

It turns out the builtin  `family()$variance` functions return the 'scaled variance' - at least sometimes. In contrast, our glmmTMB variance functions always return the variance...

An example. For the gaussian case we have:

```
y <- rnorm(100) * 10
f <- glm(y ~ 1)
sd(residuals(f, type="pearson")) ## Ca 10
```

I.e. residuals are not standardized because the variance function returns a vector of ones.
In the poisson case

```
y <- rpois(100, 10)
f <- glm(y ~ 1, family=poisson())
sd(residuals(f, type="pearson")) ## Ca 1
```

Residuals are standardized because the variance function actually *does* return the variance.

Do we really want to adopt this in `glmmTMB` ? Wouldn't it seem cleaner to use the actual variance and thereby returning standardized residuals consistently ?

**Comment by mebrooks** [comment]
I agree with @kaskr and it seems easy enough to add gaussian, Gamma, and binomial to `family.R`. Would that cause conflicts with the builtin family functions? What else would we need to change?

**Comment by bbolker** [comment]
I don't want to rush into this. We certainly shouldn't mask the built-in `gaussian()` and `Gamma()` (those are the only two we'd have to change: I don't think there are any other built-in GLM families that have a scale parameter).  As usual, a lot of the confusion comes from the distinction between the GLM context (where we only estimate the fixed dispersion parameter *after* fitting the model) and the MLE context ...

**Comment by bbolker** [comment]
Just want to point out that in general it's easier to provide the scaled variance; that way we don't have to mess around with storing additional dispersion parameter information in the environment of the function for cases where the functions are called by R infrastructure that expects to pass only a single parameter (as in negative.binomial() ...)

**Comment by bbolker** [comment]
I've been thinking about this some more.

**tl;dr** I propose that the variance function returns a value with an **attribute** "scaled" that is TRUE by default (i.e., if `is.null(attribute(m$variance(.),"scaled"))` we assume that we've got a scaled variance). If `scaled` then we have `Var=sigma(.)^2*family(.)$variance(mu)^2`; otherwise `Var=family(.)$variance(mu,phi)`. Moreover, I think we should try to store necessary dispersion parameters in the environment of the variance function when possible (cf. `MASS::negative.binomial()`/`glm.nb()`), which can be used if the `phi` parameter is missing - this will enhance compatibility with outside functions.

- scaled: all families provided by base R (gaussian, Poisson, binomial, Gamma); also beta, beta-binomial, nbinom1, genpois
- unscaled: truncated distributions, compois (?)

This also means that at least for the scaled-variance families `sigma()` should always return the necessary value to make `sigma^2*scaled_var` correct ...

**Comment by bbolker** [comment]
is there time to implement this before our July 2 deadline??
